src/shogun/modelselection/NewGridSearch.cpp

```diff
@@ -0,0 +1,891 @@
+	    tree->m_parent->ref_count())
+	auto* result = tree->m_parent->clone();
+	SG_SDEBUG("%s: %d\n", result->get_name(), result->ref_count())
+	auto result_params = result->get_params();
+	for (const auto& node : tree->m_nodes)
+	{
+		auto type_index_i =
+		    std::type_index(result_params[node.first]->get_value().type_info());
+		auto* obj = ParameterNode::to_object(node.second[0]);
+		if (type_index_i == std::type_index(typeid(CKernel*)))
+		{
+			SG_SDEBUG(
+			    "REFCOUNT - obj %s %d\n", obj->get_name(), obj->ref_count())
+			result->put(node.first, obj->as<CKernel>());
+		}
+		else if (type_index_i == std::type_index(typeid(CDistance*)))
+		{
+			SG_SDEBUG(
+			    "REFCOUNT - obj %s %d\n", obj->get_name(), obj->ref_count())
+			result->put(node.first, obj->as<CDistance>());
+		}
+		else
+		{
+			SG_SERROR(
+			    "Unsupported type %s for parameter %s::%s\n",
+			    demangled_type(type_index_i.name()).c_str(), result->get_name(),
+			    obj->get_name());
+		}
+		// obj is not needed in this scope anymore
+		SG_UNREF(obj)
+		SG_SDEBUG("REFCOUNT - obj %s %d\n", obj->get_name(), obj->ref_count())
+	}
+	std::string param_name;
+	auto put_scalar_lambda = [&result, &param_name](const auto& val) {
+		result->put(param_name, val);
+	};
+	for (const auto& param : tree->m_param_mapping)
+	{
+		param_name = param.first;
+		sg_any_dispatch(param.second, sg_all_typemap, put_scalar_lambda);
+	}
+	return result;
+}
+
+void ParameterNode::replace_node(
+    const std::string& node_name, size_t index, ParameterNode* node)
+{
+	auto it = m_nodes.find(node_name);
+
+	if (it != m_nodes.end())
+	{
+		if (index < (*it).second.size())
+			(*it).second[index].reset(node);
+		else
+		{
+			SG_SERROR(
+			    "Index out of range for %s::%s", get_parent_name(),
+			    node_name.c_str())
+		}
+	}
+	else
+	{
+		SG_SERROR(
+		    "Unable to find and replace %s::%s", get_parent_name(),
+		    node_name.c_str())
+	}
+}
+
+ParameterNode::~ParameterNode(){
+    SG_SDEBUG("~%s\n", get_name()) SG_SDEBUG(
+        "REFCOUNT %s %d\n", m_parent->get_name(), m_parent->ref_count())
+        SG_UNREF(m_parent) SG_SDEBUG(
+            "REFCOUNT %s %d\n", m_parent->get_name(), m_parent->ref_count())}
+
+ParameterNode::ParameterNode(const ParameterNode& other)
+{
+	SG_REF(other.m_parent)
+	m_parent = other.m_parent;
+	SG_SDEBUG("%s refcount: %d\n", m_parent->get_name(), m_parent->ref_count())
+	for (const auto& node : other.m_nodes)
+	{
+		for (const auto& inner_node : node.second)
+			create_node(node.first, inner_node.get());
+	}
+
+	for (const auto& param : other.m_param_mapping)
+	{
+		m_param_mapping[param.first].clone_from(
+		    other.m_param_mapping.at(param.first));
+	}
+}
+
+GridParameters::GridParameters(CSGObject* model)
+    : m_first(true), m_node_complete(false)
+{
+	SG_REF(model)
+	m_parent = model;
+	SG_SDEBUG("%s refcount: %d\n", m_parent->get_name(), m_parent->ref_count())
+	// TODO: once all hyperparameters are flagged properly replace getter with
```

Comment:
ah I see. I would do that now already, then we can have an entrance task for students to make things model-selectable ...